### PR TITLE
Feature/ak #2621 automate build process

### PR DESCRIPTION
**Pull request by abhishek1508**
## Description

Automate generate the versioning for Core and UI SDK [#2621](https://github.com/mapbox/mapbox-navigation-android/issues/2621)

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

### Goal

Automate generate the versioning for Core and UI SDK

### Implementation

Automate generate the versioning for Core and UI SDK

## Screenshots or Gifs

Please include all the media files to give some context about what's being implemented or fixed. It's not mandatory to upload screenshots or gifs, but for most of the cases it becomes really handy to get into the scope of the feature / bug being fixed and also it's REALLY useful for UI related PRs screenshot gif

## Testing

Please describe the manual tests that you ran to verify your changes

- [x] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [x] I have tested via a test drive, or a simulation/mock location app
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

**Comment by abhishek1508** [comment]
As of in `master` we need to do the following to make a release
- Cut a new branch
- Change `POM_VERSION_NAME` in all the modules (respectively depending on core or ui release)
- Commit the changes
- Allow the CI to run to make sure the checks pass
- Create a new tag as specified in the `circle.yml` off the branch created above.

This Pull Request reduces all the efforts mentioned above:
- All the admin needs to do is cut a TAG from `master` and new artifacts will be available.

### Core
Example: 
release-core-.* > release-core-1.0.0
qa-release-core-.* > qa-release-core-1.0.0
prod-release-core-.* > prod-release-core-1.0.0

The `job` checks if `CIRCLE_TAG` is `release-core-1.0.0` the following is done
```
POM_VERSION_NAME = `echo $CIRCLE_TAG | cut -d '-' -f 3`
```
The `-d` splits the `$CIRCLE_TAG` based on delimiter `-` thereby creating an array of 3 Strings.
The `-f ` specifies the `field` and `3` suggests that we are interested in the 3rd value in the array i.e. `1.0.0`

Similarly its done for all the other core releases.

### UI
Example: 
release-ui-.*-core-.* > release-ui-1.0.0-core-1.0.0
qa-release-ui-.*-core-.* > qa-release-ui-1.0.0-core-1.0.0
prod-release-ui-.*-core-.* > prod-release-ui-1.0.0-core-1.0.0

The `job` checks if `CIRCLE_TAG` is `release-ui-1.0.0-core-1.0.0` the following is done
```
POM_VERSION_NAME = `echo $CIRCLE_TAG | cut -d '-' -f 3`
```
The `-d` splits the `$CIRCLE_TAG` based on delimiter `-` thereby creating an array of 5 Strings.
The `-f ` specifies the `field` and `3` suggests that we are interested in the 3rd value in the array i.e. `1.0.0`

Similarly its done for all the other UI releases.

### SNAPSHOT

For `SNAPSHOT` versions, we don't use the `$CIRCLE_TAG` but have hardcoded the version to `1.0.0-SNAPSHOT`. This can be changed to `2.0.0-SNAPSHOT` once `1.0.0` production release is out.

### CONCLUSION
Hence, we don't need to specify `POM_VERSION_NAME` in `gradle.properties` file in any other module, since this value is stored in the `ENVIRONMENT` and `artifact-settings.gradle` reads it from the `ENVIRONMENT` using the command `System.getenv('POM_VERSION_NAME')`

**Comment by abhishek1508** [comment]
Thinking it over, I think this process fails at one point. With the way I am trying to release `UI` I never encounter for the `CORE` version associated with it. Hence, the CI releases both `UI` and `CORE` with the version specified with the `UI`.

**Comment by abhishek1508** [comment]
The latest commit will fix the issue mentioned in the [comment](https://github.com/mapbox/mapbox-navigation-android/pull/2623#issuecomment-602267041). 
I make sure that I check what kind of `SDK > Core/UI` is being released and assign values to `POM_CORE_VESION_NAME` or `POM_UI_VERSION_NAME` respectively.

File: circle.yml [hunk 1]
```diff
@@ -459,6 +500,8 @@ jobs:
     docker:
       - image: mbgl/61abee1674:android-ndk-r18
     steps:
+      - generate-core-version-name:
+          version: "1.0.0-SNAPSHOT"
```
Comment (LukasPaczos) [inline circle.yml]: As mentioned above, this is not needed. We should use the version from `gradle.properties`.

File: circle.yml [hunk 2]
```diff
@@ -478,6 +521,8 @@ jobs:
     environment:
       BINTRAY_REPO: mapbox
     steps:
+      - generate-core-version-name:
+          version: $CIRCLE_TAG
```
Comment (LukasPaczos) [inline circle.yml]: Not needed, same above and below.

File: gradle/artifact-settings.gradle [hunk 3]
```diff
@@ -10,7 +10,13 @@ ext {
     mapboxArtifactScmUrl = 'scm:git@github.com:mapbox/mapbox-navigation-android.git'
     mapboxArtifactLicenseName = 'BSD'
     mapboxArtifactLicenseUrl = 'https://opensource.org/licenses/BSD-2-Clause'
-    versionName = project.property('POM_VERSION_NAME')
+    releaseSdk = project.property('RELEASE_SDK')
+    if ( releaseSdk == 'core' ) {
+        versionName = System.getenv('POM_CORE_VERSION_NAME') ?: '1.0.0-SNAPSHOT'
+    }
+    else {
```
Comment (LukasPaczos) [inline gradle/artifact-settings.gradle]:
```suggestion
    else if ( releaseSdk == 'ui' ) {
```

File: gradle/artifact-settings.gradle [hunk 4]
```diff
@@ -10,7 +10,13 @@ ext {
     mapboxArtifactScmUrl = 'scm:git@github.com:mapbox/mapbox-navigation-android.git'
     mapboxArtifactLicenseName = 'BSD'
     mapboxArtifactLicenseUrl = 'https://opensource.org/licenses/BSD-2-Clause'
-    versionName = project.property('POM_VERSION_NAME')
+    releaseSdk = project.property('RELEASE_SDK')
+    if ( releaseSdk == 'core' ) {
+        versionName = System.getenv('POM_CORE_VERSION_NAME') ?: '1.0.0-SNAPSHOT'
```
Comment (LukasPaczos) [inline gradle/artifact-settings.gradle]:
```suggestion
        versionName = System.getenv('POM_CORE_VERSION_NAME') ?: project.property('POM_CORE_SNAPSHOT_VERSION_NAME')
```

and add `POM_CORE_SNAPSHOT_VERSION_NAME` into the root `gradle.properties` file.

File: gradle/artifact-settings.gradle [hunk 5]
```diff
@@ -10,7 +10,13 @@ ext {
...
+    }
+    else {
+        versionName = System.getenv('POM_UI_VERSION_NAME') ?: '1.0.0-SNAPSHOT'
```
Comment (LukasPaczos) [inline gradle/artifact-settings.gradle]: Same as above, but `POM_UI_SNAPSHOT_VERSION_NAME`.

File: circle.yml (hunk 1 above)
Comment (Guardiola31337) [reply circle.yml]: Great catch. I also noticed when publishing locally.

File: gradle/artifact-settings.gradle (hunk 4 above)
Comment (Guardiola31337) [reply gradle/artifact-settings.gradle]: Could we also add third condition? Something similar to https://github.com/mapbox/mapbox-navigation-android/blob/3104a0118c40d0780969b2dce13c5ac48e05d708/gradle/artifact-settings.gradle#L17 I had to add that to publish locally 👀 

```diff
diff --git a/build.gradle b/build.gradle
index a1a38d07a..7321f6529 100644
--- a/build.gradle
+++ b/build.gradle
@@ -5,6 +5,7 @@ buildscript {
   apply from: "${rootDir}/gradle/dependencies.gradle"
 
   repositories {
+    mavenLocal()
     google()
     mavenCentral()
     jcenter()
@@ -37,6 +38,7 @@ task testReport(type: TestReport, group: 'Build') {
 
 allprojects {
   repositories {
+    mavenLocal()
     google()
     mavenCentral()
     jcenter()
diff --git a/gradle.properties b/gradle.properties
index 81c91b2b9..7742c06c4 100644
--- a/gradle.properties
+++ b/gradle.properties
@@ -1,6 +1,8 @@
 # Project-wide Gradle settings.
 VERSION_NAME=0.43.0-SNAPSHOT
 GROUP=com.mapbox.mapboxsdk
+POM_CORE_VERSION_NAME=1.0.0-qa.5
+POM_UI_VERSION_NAME=1.0.0-qa.2
 
 POM_URL=https://github.com/mapbox/mapbox-navigation-android
 POM_SCM_URL=https://github.com/mapbox/mapbox-navigation-android
diff --git a/gradle/artifact-settings.gradle b/gradle/artifact-settings.gradle
index 3783c0b4e..e136a93ff 100644
--- a/gradle/artifact-settings.gradle
+++ b/gradle/artifact-settings.gradle
@@ -12,10 +12,10 @@ ext {
     mapboxArtifactLicenseUrl = 'https://opensource.org/licenses/BSD-2-Clause'
     releaseSdk = project.property('RELEASE_SDK')
     if ( releaseSdk == 'core' ) {
-        versionName = System.getenv('POM_CORE_VERSION_NAME') ?: '1.0.0-SNAPSHOT'
+        versionName = project.property('POM_CORE_VERSION_NAME') ?: '1.0.0-SNAPSHOT'
     }
     else {
-        versionName = System.getenv('POM_UI_VERSION_NAME') ?: '1.0.0-SNAPSHOT'
+        versionName = project.property('POM_UI_VERSION_NAME') ?: '1.0.0-SNAPSHOT'
     }
 
     mapboxBintrayUserOrg = 'mapbox'
```

File: circle.yml [hunk 6]
```diff
@@ -246,6 +246,47 @@ commands:
...
+  generate-ui-version-name:
+    parameters:
+      version:
+        description: version name
+        type: string
+    steps:
+      - run:
+          name: Generate UI version name
+          command: |
+            if [[ parameters.version == release-core-.*ui-.* ]]; then
```
[the remark this answers is not in the document]
Comment (abhishek1508) [reply circle.yml]: Good catch. Yeah this is a typo.

File: circle.yml [hunk 7]
```diff
@@ -246,6 +246,41 @@ commands:
...
+            if [[ $CIRCLE_TAG == release-ui-.*-core-.* ]]; then
+              POM_UI_VERSION_NAME=`echo $CIRCLE_TAG | cut -d '-' -f 3`
+              POM_CORE_VERSION_NAME=`echo $CIRCLE_TAG | cut -d '-' -f 5`
+            elif [[ $CIRCLE_TAG == qa-release-ui-.*-core-.*
+                      || $CIRCLE_TAG == prod-release-ui-.*-core-.* ]]; then
+              POM_UI_VERSION_NAME=`echo $CIRCLE_TAG | cut -d '-' -f 4`
+              POM_CORE_VERSION_NAME=`echo $CIRCLE_TAG | cut -d '-' -f 6`
+            else
+              echo "Exiting the job as this is not a release TAG"
+              circleci step halt
```
Comment (LukasPaczos) [inline circle.yml]: Looking at docs, this will skip the job, but also report success. Would `exit 1` work instead, indicating that the job failed?

Comment (abhishek1508) [reply circle.yml]: Depends. Do we completely want to stop the CI and report error or do we want to let the CI run but not publish the artifacts?

Comment (LukasPaczos) [reply circle.yml]: I'd say we want to stop. If we cut a tag that doesn't fall into the above filters, it means that there's something wrong.

Comment (LukasPaczos) [reply circle.yml]: But of course, we'll first need to check if the tag exists in the first place, like:
```
if [[ -n "$CIRCLE_TAG" ]]; then
```

File: gradle.properties [hunk 8]
```diff
@@ -14,6 +14,9 @@ POM_LICENCE_DIST=repo
 POM_DEVELOPER_ID=mapbox
 POM_DEVELOPER_NAME=Mapbox
 
+POM_UI_SNAPSHOT_VERSION_NAME=1.0.0-SNAPSHOT
+POM_CORE_SNAPSHOT_VERSION_NAME=1.0.0-SNAPSHOT
```
Comment (Guardiola31337) [inline gradle.properties]: https://github.com/mapbox/mapbox-navigation-android/pull/2623#discussion_r396428277 addressed by ☝️

Comment (Guardiola31337) [reply gradle.properties]: It'd be great to add `publish-local` entries to `Makefile` but definitely not blocking this PR, feel free to include here or in a different one 👍

Comment (abhishek1508) [reply gradle.properties]: Merging this PR. Find an issue where I am not able to run `dokka` tasks for core. Let us address this in a separate PR along with `Makefile` changes.

**Pull request merged**